Speaker Feedback: Add main feedback page

ryelle opened the pull request:
Fixes #336 — Add a page called "Leave Feedback" with a list of sessions. This will navigate to the session feedback form (TBD). The page is created on activation, and removed on deactivation, and it's otherwise not possible for most users to delete the page (it can be done through wp-cli).

### Screenshots

![Screen Shot 2020-02-06 at 6 41 01 PM](https://user-images.githubusercontent.com/541093/73988464-cd3f3f00-4910-11ea-9e49-ee05decc4ec2.png)

The page has a special label in the list table:

![Screen Shot 2020-02-06 at 7 21 17 PM](https://user-images.githubusercontent.com/541093/73990205-776d9580-4916-11ea-810d-c8878d2f1573.png)

### How to test the changes in this Pull Request:

1. The page is created on activation, so deactivate/reactivate the plugin
2. You should see a page called Leave Feedback, at site.wordcamp.org/feedback
3. This will list all sessions, and when "give feedback" is activated, navigate to that session page.

ryelle committed:
Create main Feedback page

ryelle committed:
Add basic styles & navigation functionality

ryelle committed:
Prevent the page from being deleted

ryelle committed:
Add a label to the page in the list table

ryelle committed:
Add specific class to speaker feedback nav form

coreymckrill commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/wordcamp-speaker-feedback.php:
```diff
@@ -25,4 +27,64 @@
+		'post_type'   => 'page',
+	) );
+	if ( $page_id > 0 ) {
+		update_option( 'feedback_page', $page_id );
```
The option key is hard-coded in a bunch of places. Might be better to define a constant for it.

Also, I was thinking that `feedback` is generic enough that we might run into some namespace issues with other plugins (e.g. Jetpack) if we're not careful. In some other places I see you're using `sft_` as a prefix. Could be worth doing for the option key too.

coreymckrill commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/wordcamp-speaker-feedback.php:
```diff
@@ -25,4 +27,64 @@
+/**
+ * Create main Feedback page.
+ */
+function activate() {
+	add_feedback_page();
+}
```
I'm wondering if this will work if this plugin ends up being network activated. I.e., when a new site is created, and this is already network activated, will it still run this and create the page on the new site?

Unfortunately, when I tried to test this on my Docker, I kept running into memory limits and max execution times, which is probably (hopefully) an unrelated problem, so I don't know the answer yet. I did notice that when I network activated the plugin, existing sites in the network did not have a feedback page added.

ryelle replied:
Good point, I didn't test this network-wide.

ryelle replied:
I've fixed it so that it will also check for the page on `init` and add it if necessary… do you think I should also update the deactivation cleanup to account for network deactivations? seems like that would probably time out on wc.org, looping through every site just to delete a page.

coreymckrill replied:
> do you think I should also update the deactivation cleanup to account for network deactivations? seems like that would probably time out on wc.org, looping through every site just to delete a page.

Yeah, probably not worth it to worry about deactivation. This did make me think that we should probably add a skip feature flag here though, and only add the page on sites past a certain ID.

ryelle replied:
> skip feature flag

should the entire plugin functionality be on a skip feature flag? or would we still want the `/session/blah/feedback` form to work on every site?

coreymckrill replied:
Good point, it should probably just be the entire plugin.

ryelle replied:
We can do that in a separate PR, then 👍

coreymckrill commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/wordcamp-speaker-feedback.php:
```diff
@@ -25,4 +27,64 @@
+ * Create the Feedback page, save ID into an option.
+ */
+function add_feedback_page() {
+	$page_id = wp_insert_post( array(
```
Since the content is getting overwritten by a filter, I wonder if it would be helpful to insert an admin note into the post content here explaining how this page works.

coreymckrill commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/page.php:
```diff
@@ -0,0 +1,45 @@
+	$feedback_page = (int) get_option( 'feedback_page' );
+
+	if ( $feedback_page === $post->ID ) {
+		$post_states['sft-page'] = _x( 'Speaker Feedback', 'wordcamporg' );
```
Missing the context parameter in the `_x()` here.

coreymckrill commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/views/form-select-sessions.php:
```diff
@@ -0,0 +1,36 @@
+						esc_attr( get_the_ID() ),
+						wp_kses_post( get_the_title() )
+					);
+				} ?>
```
Looks like it's not actually required in the WP [coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#brace-style), but I generally think it's a good idea to use the alternative control structure syntax when mixing HTML and PHP.

ryelle committed:
Use a constant for the feedback page option name

ryelle committed:
PR feedback

Add the page to the site if it doesn't exist; add content to the page

ryelle committed:
Use alternate control structure

coreymckrill commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/wordcamp-speaker-feedback.php:
```diff
@@ -16,13 +16,92 @@
+function deactivate() {
+	$page_id = get_option( OPTION_KEY );
+	wp_delete_post( $page_id, true );
+}
```
Need to delete the option here as well, since the page is being permanently deleted.

coreymckrill commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/wordcamp-speaker-feedback.php:
```diff
@@ -16,13 +16,92 @@
+ * Create the Feedback page, save ID into an option.
+ */
+function add_feedback_page() {
+	$page_id = get_option( OPTION_KEY );
+	if ( $page_id ) {
+		return;
+	}
```
It might be a good idea check and make sure an actual page is attached to this ID. Although, if the option gets deleted when the plugin is deactivated (per my other suggestion), that might eliminate any edge cases where the option still exists but the page doesn't...

I guess if this is running on every page load on `init` having the extra query for a specific page isn't great...

ryelle replied:
I've updated it so that when the page is deleted, the option is deleted too — so we _shouldn't_ run into a case where the option still exists for a missing page.

ryelle committed:
Delete the saved option when the page is deleted

ryelle commented:
@coreymckrill I think I've addressed all your feedback here ~is this ready to merge now?~ Edit: going to merge this since you're AFK, but can address any issues in following PRs 👍
